File: l5-ctci/backtracking/47.java
import java.util.*;

class Solution { // avoid duplicates by sorting nums and skipping the same numbers

    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        backtrack(nums, new boolean[nums.length], new ArrayList<>());
        return res;
    }

    private void backtrack(int[] nums, boolean[] usedInd, List<Integer> path) {
        if (path.size() == nums.length) {
            res.add(new LinkedList<>(path));
            return;
        }
        int lastUsed = Integer.MIN_VALUE;
        for (int i = 0; i < usedInd.length; i++) {
            if (!usedInd[i] && nums[i] != lastUsed) {
                lastUsed = nums[i];
                path.add(nums[i]);
                usedInd[i] = true;
                backtrack(nums, usedInd, path);
                usedInd[i] = false;
                path.remove(path.size() - 1);
            }
        }
    }
}

class SolutionHashing { // avoid duplicates by using only unique numbers as an option (create hashmap with counter)

    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> permuteUnique(int[] nums) {
        Map<Integer, Integer> counter = new HashMap<>();
        for (var n : nums) {
            counter.put(n, counter.getOrDefault(n, 0) + 1);
        }
        backtrack(nums, counter, new ArrayList<>());
        return res;
    }

    private void backtrack(int[] nums, Map<Integer, Integer> counter, List<Integer> path) {
        if (path.size() == nums.length) {
            res.add(new LinkedList<>(path));
            return;
        }
        for (var num : counter.keySet()) {
            if (counter.get(num) == 0) {
                continue;
            }
            path.add(num);
            counter.put(num, counter.get(num) - 1);
            backtrack(nums, counter, path);
            counter.put(num, counter.get(num) + 1);
            path.remove(path.size() - 1);
        }
    }
}

File: repeat_py/47.py
from typing import List


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res: List[List[int]] = list()

        counts = dict()
        for n in nums:
            if n not in counts:
                counts[n] = 1
            else:
                counts[n] += 1

        def permRec(counts, current):
            if len(current) == len(nums):
                res.append(current.copy())
                return

            for val in counts:
                if counts[val]:
                    current.append(val)
                    counts[val] = counts[val] - 1
                    permRec(counts, current)
                    current.pop()
                    counts[val] = counts[val] + 1

        permRec(counts, list())

        return res


Solution().permuteUnique([1,1,2])
